### PR TITLE
Added support for generic push buttons

### PR DESCRIPTION
**Pull request by maximweb**
I've added an initial draft for the support of generic push buttons.

The library includes two types of software debounce:
* A minimum time a button is allowed to be repressed.
* A minimum debounce time the button has to be pressed.

_Note_: While I use a button with pullup resistor as hardware, I also added an RC low pass filter and an Schmitt trigger. The latter inverts the voltage signal. So if you want to test without Schmitt trigger, you might need to replace `RISING` \[[1](https://github.com/brewpeople/brewslave/blob/push-buttons/PushButton.cpp#L16),[2](https://github.com/brewpeople/brewslave/blob/push-buttons/app.cpp#L232-L237)\] and `HIGH` \[[3](https://github.com/brewpeople/brewslave/blob/push-buttons/PushButton.cpp#L34)\] with `FALLING` and `LOW`, respectively. We might want to make this configurable?

@matze, I wanted to wrap the Arduino-specific `attachInterrupt` call within the button's `begin` function. Unfortunately `attachInterrupt` requires a static function to be passed, and I do not have the faintest idea how to do this.

Calling `attachInterrupt(pin, button_instance.trigger, RISING)` resulted in compile error `is a non-static member...`

So I tried to define [a wrapper](https://github.com/brewpeople/brewslave/blob/push-buttons/app.cpp#L45) and pass it to the button instance:
```
void brew_button_trigger();          // forward declaration
PushButton brew_button{BREW_BUTTON_PIN, brew_button_trigger};
void brew_button_trigger() {
    brew_button.trigger();
}
```
and [call it within `begin()`](https://github.com/brewpeople/brewslave/blob/push-buttons/app.cpp#L230-L231)
```
void setup()
{
    brew_button.begin(); // calling the `attachInterrupt`
}
```

This way I get tons of invalid type errors, I tried to adapt stuff from [the original `attachInterrupt` implementation](https://github.com/arduino/ArduinoCore-avr/blob/9f8d27f09f3bbd1da1374b5549a82bda55d45d44/cores/arduino/WInterrupts.c), but failed. 
The respective lines are commented out. _Can you help?_

### CODE REVIEW

**Comment by matze** [comment]
> Unfortunately attachInterrupt requires a static function to be passed, and I do not have the faintest idea how to do this.

A static function is any ordinary free function or a method declared `static`. Hence, it's impossible to attach any member method. So you have to do it like the encoder via ISR or by attaching a free function and calling the method. But I don't see the need to abstract it anyhow. Being called from outside you get the same behavior without re-implementing the same shit over and over.

P.S.: I start to see trailing whitespace again (maybe change your editor to show them) and inconsistent braces.

**Comment by maximweb** [comment]
@matze Your clang workflow does not provide any user-friendly returns...

I removed the remains from development, formatted according to clang file, and rebased with your reformatted branch #26 .

**Comment by matze** [comment]
I merged, so rebase on top of master and you won't have my commits in the PR.

File: app.cpp [hunk 1]
```diff
@@ -160,6 +188,11 @@ class App {
             m_ui_state &= ~Ui::State::Warning;
         }
 
+        brew_button.update();
+        if (brew_button.pressed()) {
+            // do stuff
+        }
```
Comment (matze) [inline app.cpp:194]: Remove this until we really do implement logic.

**Review by matze — APPROVED** [review]
Besides that minor nit, approved.

**Pull request merged**